cmd/fs-v1.go

```diff
@@ -268,6 +268,58 @@ func (fs *FSObjects) StorageInfo(ctx context.Context) StorageInfo {
 	return storageInfo
 }
 
+// Update the data usage while walking recursively through the given bucket/prefix
+func (fs *FSObjects) walkDataUsageInfo(ctx context.Context, info *DataUsageInfo, bucket, prefix string) {
+	entries, err := readDir(pathJoin(fs.fsPath, bucket, prefix))
+	if err != nil {
+		return
+	}
+
+	for _, entry := range entries {
+		if strings.HasSuffix(entry, "/") {
+			fs.walkDataUsageInfo(ctx, info, bucket, pathJoin(prefix, entry))
+		} else {
+			info.ObjectsCount++
+
+			// Stat the file to get file size.
+			fi, err := fsStatFile(ctx, pathJoin(fs.fsPath, bucket, prefix, entry))
+			if err != nil {
+				continue
+			}
+			size := uint64(fi.Size())
+			info.ObjectsTotalSize += size
+			info.BucketsSizes[bucket] += size
+			info.ObjectsSizesHistogram[objSizeToHistoInterval(size)]++
+		}
+	}
+}
+
+// DataUsageInfo returns data usage stats of the current FS deployment
+func (fs *FSObjects) DataUsageInfo(ctx context.Context) DataUsageInfo {
+	buckets, err := readDir(fs.fsPath)
+	if err != nil {
+		return DataUsageInfo{}
+	}
+
+	var info = DataUsageInfo{
+		BucketsCount:          uint64(len(buckets)),
+		BucketsSizes:          make(map[string]uint64),
+		ObjectsSizesHistogram: make(map[string]uint64),
+	}
+
+	for _, bucket := range buckets {
+		bucket = strings.TrimSuffix(bucket, "/")
+		// Ignore all reserved bucket names and invalid bucket names.
+		if isReservedOrInvalidBucket(bucket, false) {
+			continue
+		}
+		fs.walkDataUsageInfo(ctx, &info, bucket, "")
```

Comment:
You should do a more memory efficient implementation here @vadmeste this looks like can easily run-out memory for FS deployments

Reply:
I am not sure if you noticed that I pass info as a pointer that will be modified by further recursive walkDataUsageInfo() calls. There is actually only one data usage info.

I think the only thing that can consume memory is the return list of readDir() function.

Reply:
The problem is that it's not streaming, its not discarding the info quickly but storing it in memory all along until top layer uses it to do some accounting. I think it's better to do accounting directly at posix level and top level layer only does some form of counting.

Reply:
> its not discarding the info quickly but storing it in memory all along until top layer uses it to do some accounting.

Yes, but in the whole ongoing process, there is only one struct which is being passed recursively (you can see I passed a pointer), and in each recursive call, the same struct is being updated.

Reply:
The problem is this implementation is going to be very slow. Try exploring the fastwalk approach which will allow for faster scans.

Perhaps we can avoid map and instead use a slice of bucket names to capture this information. StatVol can return this information along with it to the top layer for consuming it.